src/redis-cli.c

```diff
@@ -798,6 +740,165 @@ void cliInitCommandHelpEntries(redisReply *commandTable, dict *groups) {
     }
 }
 
+/* Does the server version support a command/argument only available "since" some version?
+ * Returns 1 when supported, or 0 when the "since" version is newer than "version". */
+static int versionIsSupported(sds version, sds since) {
+    int i;
+    char *versionPos = version;
+    char *sincePos = since;
+    if (!since) {
+        return 1;
+    }
+
+    for (i = 0; i != 3; i++) {
+        int versionPart = atoi(versionPos);
+        int sincePart = atoi(sincePos);
+        if (versionPart > sincePart) {
+            return 1;
+        } else if (sincePart > versionPart) {
+            return 0;
+        }
+        versionPos = strchr(versionPos, '.');
+        sincePos = strchr(sincePos, '.');
+        if (!versionPos || !sincePos)
+            return 0;
+        versionPos++;
+        sincePos++;
+    }
+    return 0;
+}
+
+static void removeUnsupportedArgs(struct cliCommandArg *args, int *numargs, sds version) {
+    int i = 0, j;
+    while (i != *numargs) {
+        if (versionIsSupported(version, args[i].since)) {
+            i++;
+
+            if (args[i].subargs) {
+                removeUnsupportedArgs(args[i].subargs, &args[i].numsubargs, version);
+            }
+            continue;
+        }
+        for (j = i; j != *numargs; j++) {
+            args[j] = args[j + 1];
+        }
+        (*numargs)--;
+    }
+}
```

Comment:
@jhelbaum maybe you can help me find the bug here (before i dive into the code to figure it out).

this crashes when redis-cli connects to redis 6.2 (setting up the legacy help).
it happens on CLUSTER SLAVES, which has `numsubargs` set to 0.
but it advances `i++` before accessing `args[i].subargs`.

Reply:
I traced it to the same section of code too using dbg. It's a recursive call to removeUnsupportedArgs.

What's slightly odd here is that this code modifies the global command table directly, at least for subargs.

Reply:
@oranagra I think you just solved the problem in your comment.

> it advances `i++` before accessing `args[i].subargs`

Moving `i++` to after accessing `args[i].subargs` (to just before `continue`) seems to solve the problem.

Reply:
Just looking at the code, that makes sense. It's hard to understand how it ever worked in the first place, and I'm sure I tested it manually.

It seemed too complicated to set up an automated test, with the need to launch a server and a client running different versions of the code.

I'd recommend exploring the behavior on a few commands with subcommands to make sure it looks right.

Reply:
maybe we can extend the tests to run in 3 modes:
1. checking all completion hints when they consume the info form commands.c (when redis-cli is unable to connect to a server)
2. checking all completion hints when they consume the info form the latest COMMAND DOCS (setting up a redis server of the latest version for that).
3. checking part of the completion hints (skipping some) when using just COMMAND command (not COMMAND DOCS), as if we're connected to an older version that doesn't support it.

WDYT?